ynput/ayon-launcher
# Shims for AYON launcher executables

## Changelog Description
Implemented simple shims which can redirect to ayon launcher version.

## Additional info
Shims are meant to help with registering of path to AYON launcher. Right now you can point to a single version of AYON launcher, which would stop to work when the version is uninstalled because new version is available. Only think which shim does is to find an existing ayon launcher and execute it, it does not download it or handle logic.

The downside of this approach is that it does not handle cases when there are any available versions. In that case it just crashes. It also does not autodetect available versions, so the new launcher must be launched at least once so shim can find it.

The shim is versioned with `version` file which contains version raw string. This file was used instead of `version.py` because of potential future change of implementation from Python to C++.

AYON launcher has new argument handling. When `"init-ayon-launcher"` is passed it just stores information about executable and deploys shims witout any other action.

## Full feature set of this PR
- AYON launcher has shim, which is capable of finding real AYON launcher executable. That is executable to which can be pointed any automation which should support any version of AYON launcher (e.g. deadline Ayon plugin).
- The shim is also used for custom `ayon-launcher://` protocol scheme, that is different based on OS. With this support we can start to use webactions.
- Login window can force username to be logged in.

## Warning
On windows and linux is shim launching new ayon launcher and the shim process ends after the AYON launcher process ends and is using the same exit code and is passing in stdout and stderr (if is available). That is not true on macOs where shim only opens new launcher and ends.

Reason: Shim is used for webactions, api of uri handling on macOs is based on events (launch process and send event to start uri). The issue is that only one process of the application can be running to be able to achieve this. Any future process just won't start.

Possible issues: If the exectable is used in a automation that should wait for the process to "finish", it is not possible.

## Testing notes:
1. Run Create env command.
2. Run build and make installer.
3. Install AYON launcher on windows. Start unpacked AYON launcher at least once on linux and mac.
4. On Windows It should create shims on your machine (`%LOCALAPPDATA%\Ynput\AYON\shim\`).
5. For other platforms the shims are created when the AYON launcher is launched for a first time.
6. When you run the shim it should start AYON launcher.

If you're retesting this PR, make sure to reinstall venv and remove existing AYON launcher. Also remove existing shim, `%LOCALAPPDATA%\Ynput\AYON\shim\` on windows `/Applications/AYON.app` on macOs and `~/.local/AYON/shim` on linux.


### Webactions
Shims added option to use webactions. For that make sure you have
- AYON server at least 1.3.1 (the higher the better).
- this PR
- ayon-core (develop)
Follow testing notes from https://github.com/ynput/ayon-launcher/pull/121
- ayon-applications https://github.com/ynput/ayon-applications/pull/1
